libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
+			rc.md.ExtNameServers = append(rc.md.ExtNameServers, ExtDNSEntry{
+				Addr:         addr,
+				HostLoopback: addr.IsLoopback() && !rc.md.NSOverride,
+			})
+		}
+	}
+	rc.nameServers = newNSs
+
+	// If there are no external nameservers, and the only nameserver left is the
+	// internal resolver, use the defaults as ext nameservers.
+	if len(rc.md.ExtNameServers) == 0 && len(rc.nameServers) == 1 {
+		log.G(context.TODO()).Info("No non-localhost DNS nameservers are left in resolv.conf. Using default external servers")
+		for _, addr := range defaultNSAddrs(keepIPv6) {
+			rc.md.ExtNameServers = append(rc.md.ExtNameServers, ExtDNSEntry{Addr: addr})
+		}
+		rc.md.UsedDefaultNS = true
+	}
+
+	// Validate the ndots option from host config or overrides, if present.
+	// TODO(robmry) - pre-existing behaviour, but ...
+	//   Validating ndots from an override is good, but not-liking something in the
+	//   host's resolv.conf isn't a reason to fail - just remove? (And it'll be
+	//   replaced by the value in reqdOptions, if given.)
+	if ndots, exists := rc.Option("ndots"); exists {
+		if n, err := strconv.Atoi(ndots); err != nil || n < 0 {
+			return errdefs.InvalidParameter(
+				fmt.Errorf("invalid number for ndots option: %v", ndots))
+		}
+	}
+	// For each option required by the nameserver, add it if not already
+	// present (if the option already has a value don't change it).
+	for _, opt := range reqdOptions {
+		optName, _, _ := strings.Cut(opt, ":")
+		if _, exists := rc.Option(optName); !exists {
+			rc.AddOption(opt)
+		}
+	}
+
+	rc.md.Transform = "internal resolver"
+	return nil
+}
+
+// Generate returns content suitable for writing to a resolv.conf file. If comments
+// is true, the file will include header information if supplied, and a trailing
+// comment that describes how the file was constructed and lists external resolvers.
+func (rc *ResolvConf) Generate(comments bool) ([]byte, error) {
+	s := struct {
+		Md             *metadata
+		NameServers    []string
+		ExtNameServers []string
+		Search         string
+		Options        string
+		Other          string
+		Overrides      []string
+		Comments       bool
+	}{
+		Md:       &rc.md,
+		Search:   strings.Join(rc.search, " "),
+		Options:  strings.Join(rc.options, " "),
+		Other:    rc.other,
+		Comments: comments,
+	}
+	for _, addr := range rc.nameServers {
+		s.NameServers = append(s.NameServers, addr.String())
+	}
+	for _, ns := range rc.md.ExtNameServers {
+		var nss string
+		if ns.HostLoopback {
+			nss = "host(" + ns.Addr.String() + ")"
+		} else {
+			nss = ns.Addr.String()
+		}
+		s.ExtNameServers = append(s.ExtNameServers, nss)
+	}
+	if rc.md.NSOverride {
+		s.Overrides = append(s.Overrides, "nameservers")
+	}
+	if rc.md.SearchOverride {
+		s.Overrides = append(s.Overrides, "search")
+	}
+	if rc.md.OptionsOverride {
+		s.Overrides = append(s.Overrides, "options")
+	}
+
+	templText := `{{if .Comments}}{{with .Md.HeaderComment}}{{.}}
```

Comment:
Done ... although I'm not really sure it's an improvement, now it's easier to miss changes to members of `ResolvConf` that are used by the template (because they're not searchable-for in the same file), and those don't show up until the template's used.